src/helpers/cbv.tsx

```diff
@@ -0,0 +1,140 @@
+import reducerRegistry from '@onaio/redux-reducer-registry';
+import React, { useEffect } from 'react';
+import { connect } from 'react-redux';
+import { Store } from 'redux';
+import { displayError } from '../helpers/errors';
+import { getURL } from '../services/opensrp';
+import reducer, {
+  reducerName,
+  selectAllMessages,
+  sendMessage,
+} from '../store/tests/ducks/messages';
+
+reducerRegistry.register(reducerName, reducer);
+
+/** interface to describe ObjectList options */
+interface ObjectListOptions {
+  listPropName: string;
+  dispatchPropName: string;
+}
+
+/** interface for the props of the connected component created by ObjectList  */
+interface ObjectListProps {
+  actionCreator: any;
+  objectList: any;
+}
+
+/**
+ * ObjectList class
+ *
+ * This class is initialized with the following parameters:
+ *    - component: a react component
+ *    - options: an object representing options for ObjectList
+ *
+ * The class's render method will then return the same component wrapped in a
+ * higher order component that has been connected to the redux store, and which
+ * knows how to fetch data for display.
+ *
+ * The end goal is a quick and convenient way to display a list of objects without
+ * having to worry about boilerplate code that deal with fetching objects and
+ * storing them in a Redux store.
+ *
+ * Every method in this class can and should be overridden to cater to custom needs.
+ */
+export class ObjectList {
```

Comment:
@cKellyDesign @p-netm this is a first attempt at solving part of https://github.com/onaio/js-tools/issues/126

We have a basic working version, but I need critical eyes looking at it.

- Does it meet the needs that we stated in that issue?
- What can we improve?
- Is it easy and natural to use?
- What would you do differently?

PS - it is definitely not done, and the final version will be a package in js-tools.

Reply:
- What would i do differently
I would probably keep data fetching in the dumb component. I am thinking of places where we have more than a single fetch data request, places where the request is part of the component's lifecycle or as a result of a user interaction.

i understand we can override the current `fetchData` . I at the moment i can only speculate that this would result in a bit of added complexity.

**How would i work around this**
data fetching would remain colocated with dumb component as it is now.
The dumb component would thus need to be given the necessary actionCreators, that it needs

Reply:
@p-netm I thought about that a bit.  The way I thought about it was that the process of fetching the data goes like this:

- call some URL in a very specific way (this varies a lot between components)
- while the URL is loading, display a loading indicator
- once done, show the data

I think we should DRY this up and standardize it, and by "this" I mean the whole process.  i.e. you may want to call a specific URL in a specific way which is cool, but we want to standardize what happens while you call it and after.

Thoughts on that?

Reply:
> I would probably keep data fetching in the dumb component.

@p-netm and @moshthepitt - I agree and disagree with this, I think there are two cases in which we are calling the API:
1. When a component mounts (everything on inital mounting)
2. After a component mounts (later lifecycle methods, user interaction, etc)
In my opinion, `2` should definitely remain in the dumb components while `1` should be the responsibility of the `cbv`.

To accomplish this while keeping things modular and DRY, could there be a way to pass options into `cbv` which tell it:
* which API endpoints to to call on behalf of the dumb component
* what options should be used when making the call
* what to do when the call is completed
* what to do if the call fails

If we pass an array of such options, then the `cbv` can generate a list of Promises to `resolve` or `reject`. With this approach it seems the `cbv` could also manage an `isLoading` state which is passed into the dumb component props.

What do you think of this idea?

Reply:
What gets tricky is the business logic that often comes with the pages, like if the results from request A are required to configure request B.

Reply:
@p-netm @cKellyDesign okay so I'll leave this out for now.  Feels like we would be trying to make something DRY before we fully understand how we will be using this.  We'll revisit this when the usage patterns and opportunities for optimization become clearer.

Reply:
I am satisfied with this decision